tracer/src/Datadog.Trace/ClrProfiler/AutoInstrumentation/AspNetCore/AspNetCoreBlockMiddlewareIntegration.cs

```diff
@@ -0,0 +1,109 @@
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Http.Features;
+using Microsoft.Extensions.DependencyInjection;
+#endif
+using System;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using Datadog.Trace.AppSec;
+using Datadog.Trace.ClrProfiler.CallTarget;
+using Datadog.Trace.Configuration;
+using Datadog.Trace.DuckTyping;
+using Datadog.Trace.Logging;
+
+namespace Datadog.Trace.ClrProfiler.AutoInstrumentation.AspNetCore
+{
+    /// <summary>
+    /// The ASP.NET Core middleware integration.
+    /// </summary>
+    [InstrumentMethod(
+        AssemblyName = "Microsoft.AspNetCore.Hosting",
+        TypeName = "Microsoft.AspNetCore.Hosting.Builder.ApplicationBuilderFactory",
+        MethodName = "CreateBuilder",
+        ParameterTypeNames = new[] { "Microsoft.AspNetCore.Http.Features.IFeatureCollection" },
+        ReturnTypeName = "Microsoft.AspNetCore.Builder.IApplicationBuilder",
+        MinimumVersion = "2",
+        MaximumVersion = "6",
+        IntegrationName = nameof(IntegrationId.AspNetCore))]
+    public class AspNetCoreBlockMiddlewareIntegration
+    {
+        private static readonly IDatadogLogger Log = DatadogLogging.GetLoggerFor(typeof(AspNetCoreBlockMiddlewareIntegration));
+
+        /// <summary>
+        /// test
+        /// </summary>
+        /// <param name="instance">instance</param>
+        /// <param name="returnValue">returnValue</param>
+        /// <param name="exception">exception</param>
+        /// <param name="state">state</param>
+        /// <typeparam name="TTarget">TTarget</typeparam>
+        /// <typeparam name="TReturn">TReturn</typeparam>
+        /// <returns>CallTargetReturn</returns>
+        public static CallTargetReturn<TReturn> OnMethodEnd<TTarget, TReturn>(TTarget instance, TReturn returnValue, Exception exception, in CallTargetState state)
+        {
+#if !NETFRAMEWORK
+            if (Security.Instance.Settings.Enabled)
+            {
+                var appb = (IApplicationBuilder)returnValue;
+                appb.MapWhen(context => context.Items["block"] is true, HandleBranch);
+            }
+#endif
+
+            return new CallTargetReturn<TReturn>(returnValue);
+        }
+
+#if !NETFRAMEWORK
+        internal static void HandleBranch(Microsoft.AspNetCore.Builder.IApplicationBuilder app)
+        {
+            app.Run(
+                context =>
+                {
+                    var sec = Security.Instance;
+                    var settings = sec.Settings;
+                    var httpResponse = context.Response;
+                    httpResponse.Clear();
+
+                    foreach (var cookie in context.Request.Cookies)
+                    {
+                        httpResponse.Cookies.Delete(cookie.Key);
+                    }
+
+                    // this should always be true for core, but it would seem foolish to ignore it, as potential source of future bugs
+                    if (sec.CanAccessHeaders())
```

Comment:
AFAICT, this check _can't_ be anything but true, and IMO it's not worth checking, because the conditions are behind the `#if`. What we _should_ be checking is whether the response has already started to be sent using `HasStarted`. If it _has_, I'm not sure what we should do 🤔 Potentially close the `Body` stream? We can't send a new response AFAIK, or change any headers.

Shouldn't happen but probably better to be safe?

```suggestion
                    if (httpResponse.HasStarted)
```